zebra-state/src/service/read.rs

```diff
@@ -338,62 +382,121 @@ where
     // But we can compensate for deleted UTXOs by applying the overlapping non-finalized UTXO changes.
 
     // Check if the finalized and non-finalized states match or overlap
-    let required_min_chain_root = finalized_tip_range.start().0 + 1;
-    let mut required_chain_overlap = required_min_chain_root..=finalized_tip_range.end().0;
+    let required_min_non_finalized_root = finalized_tip_range.start().0 + 1;
+
+    // Work out if we need to compensate for finalized query results from multiple heights:
+    // - Ok contains the finalized tip height (no need to compensate)
+    // - Err contains the required non-finalized chain overlap
+    let finalized_tip_status = required_min_non_finalized_root..=finalized_tip_range.end().0;
+    let finalized_tip_status = if finalized_tip_status.is_empty() {
+        let finalized_tip_height = *finalized_tip_range.end();
+        Ok(finalized_tip_height)
+    } else {
+        let required_non_finalized_overlap = finalized_tip_status;
+        Err(required_non_finalized_overlap)
+    };
```

Comment:
Previously, the code used the `required_non_finalized_overlap` range when the queries were consistent.